### ISSUE TITLE
Pattern zoom on one knob gives unexpected results

### ISSUE THREAD

**Issue by riban-bw** [zynthian/zynthian-issue-tracking#1065]
# Describe the issue
Both vertical and horizontal zoom of patterns is controlled by a single knob (2). This gives odd behaviour. When you zoom in, both axes zoom but as you zoom out, the horizontal axis reaches its maximum and stops zooming whilst the vertical axis continues to zoom. This results in the view changine shape / size in an unexpected and probably undesirable way.
# To Reproduce
## Steps to reproduce the behavior:
1. Go to pattern editor
2. Adjust knob 2 clockwise to increase zoom
3. Adjust knob 2 anticlockwise to reduce zoom
 ## Expected behaviour
Display ends in same aspect ratio as when started.
## Actual behaviour
Vertical zoom continues to decrease.
# Additional context
I appreciate the desire to reduce the quantity of knobs used for this function but the user experience is odd and possibly undesirable. It is not immediately obvious how to return the vertical zoom to its previous value without adjusting horizontal zoom. Also, I don't think zooming both axes makes much sense on a single knob. A user is likely to want to zoom in on something in context. They are either looking for specific notes (vertical) or time (horizontal) but rarely both.

# Configuration
## Hardware
```
Raspberry Pi 4 Model B Rev 1.5
Audio: V5 ADAC
Display: MIPI DSI 800x480 (inverted)
Wiring: V5
I2C: MCP23017@0x20, MCP23017@0x21
Profile: v5
```
## System
```
Debian GNU/Linux 12 (bookworm)
Build Date: 2024-02-01
Memory: 16% (613M/3792M)
SD Card: 86% (24G/29G)
Temperature: 45.7ºC
Overclock: Maximum
```
## MIDI & UI
```
Tuning: 440 Hz
Master Channel: 16
Preload Presets: On
ZS3 (SubSnapShots): On
Power Save: 10 minutes
Audio Levels on Snapshots: On
```
## Software
```
zyncoder: oram (a7976d8) 
zynthian-ui: oram (199670b)
zynthian-sys: oram (d542728)
zynthian-data: oram (db0355a)
zynthian-webconf: oram (c27fea2)
```

**Comment by riban-bw** [comment]
I have played some more with this and cannot find a way to zoom only one axis. I had assumed there would be a menu option or some mechanism but cannot find that. I really think we need a way to zoom each axis individually.

With the touchscreen, we could use similar method as audio player and use horizontal / vertical pinch zoom and two finger drag.

**Comment by riban-bw** [comment]
I have considered this a little more and have this concept of workflow:

The purpose of zoom is to provide:
- High-level view of more data (more notes, more steps)
- Access to individual cells for manipulation (finger sized note & step)

This may mean that zoom in should always get to a point where it zooms horizontal and vertical together, so as to provide the finger-sized cells. Zoom out may reach the horizontal limit first and then continue to zoom vertiacally to show more notes. (Ideally it should allow full zoom to whole 128 note key range.)

If this workflow concept is correct then I suggest zoom work thus:

- One control for both vertical and horizontal zoom (knob / pinch)
- Vertical and horizontal zoom occur from full zoom in, up to the threshold that whole pattern is shown, after which only vertical zoom out continues.
- When zooming in, only vertical zoom occurs until the threshold is reached that provides the appropriate x/y aspect ratio for each note/step for dual zoom

Maybe that is all too wordy. There is a threshold at which zoom changes from only vertical to both. That threshold is defined by the desired apsect ration of the notes and steps (maybe square or wide (not tall) rectangle) to allow reliable touch when sufficiently zoomed.

This is similar to the behaviour now except zoom in currently always zooms both. That needs to change to occur only from the threshold.

**Comment by riban-bw** [comment]
We should also be zooming around the selected cell. This does not seem to be happening. As you zoom the selected cell can be zoomed out of view.

**Comment by jofemodo** [comment]
Having separated vertical and horizontal zoom was tested and discharged because resulting workflow is not agile. Less is more, specially in this case!! Anyway, i agree that we can improve the current one-encoder zoom behavior. Let's do it as you say in your last proposal.

**Comment by jofemodo** [comment]
Now it works much better ;-)
